Add exits to yjit stats, convert at-exit reporting to Ruby

noahgibbs opened the pull request:
As discussed in https://github.com/Shopify/yjit/issues/107, this adds exit_op_count and inlined/outlined code to the returned YJIT stats. It also converts the existing exit stats report, with no changes, from C to Ruby.

I'm initially marking this as a draft because I'm seeing some small differences in the statistics themselves (e.g. number of side-exits from opt_send_without_block). Investigating now.

maximecb commented on yjit.rb:
```diff
@@ -125,7 +125,7 @@ def self.disasm_block(cs, block, highlight)
   def self.runtime_stats
     # defined in yjit_iface.c
-    Primitive.get_stat_counters
+    Primitive.get_yjit_stats
```
Good call renaming this.

maximecb commented on yjit_iface.c:
```diff
@@ -747,6 +760,22 @@ get_stat_counters(rb_execution_context_t *ec, VALUE self)
     }
+
+    {
+        // Iterate through exit_op_count
```
I think this needs a bit more commenting explaining what this is/does.

maximecb commented:
Yes that seems like an improvement 👍

noahgibbs commented:
I've changed the strcpy of a static string to strncpy to satisfy caution-tape-bot. I had fixed the potential variable-length strcpy earlier, but I think it still doesn't like me copying "exit_" with plain strcpy.

jhawthorn commented on yjit_iface.c:
```diff
@@ -747,6 +760,23 @@ get_stat_counters(rb_execution_context_t *ec, VALUE self)
+        strncpy(key_string, "exit_", 6);               // Copy 6 characters to include the NUL.
+        for (int i = 0; i < VM_INSTRUCTION_SIZE; i++) {
+            const char *i_name = insn_name(i); // Look up Ruby's NUL-terminated insn name string
+            strncpy(key_string + 5, i_name, rb_vm_max_insn_name_size + 1);
```
This looks right but might be simpler as `snprintf`, to avoid counting lengths by hand.

noahgibbs replied:
That's a good idea. Done.

maximecb commented:
@noahgibbs is this PR ready to merge?
